### PR TITLE
chore(asm): migrate waf data to asm context

### PR DESCRIPTION
**Pull request by christophe-papazian**
- use the asm context as the main tool to store data sent to the WAF
- data are still kept in the root span context for the backend
- fix a previous non impacting bug by only setting asm data for ingoing web request
- improve the set_http_meta test accordingly

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines) are followed.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] PR description includes explicit acknowledgement/acceptance of the performance implications of this PR as reported in the benchmarks PR comment.

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.

### CODE REVIEW

File: ddtrace/appsec/_asm_request_context.py [hunk 1]
```diff
@@ -94,6 +107,14 @@ def set_value(category, address, value):  # type: (str, str, Any) -> None
         asm_context_attr[address] = value
 
 
+def set_waf_address(address, value, span):  # type: (str, Any, Any) -> None
+    from ddtrace.internal._context import set_item
```
Comment (gnufede) [inline ddtrace/appsec/_asm_request_context.py:111]: Can't we move this import to the top?

**Review by gnufede — DISMISSED** [review]

File: ddtrace/appsec/_asm_request_context.py (hunk 1 above)
Comment (christophe-papazian) [reply ddtrace/appsec/_asm_request_context.py:111]: Yes we can!

**Review by gnufede — DISMISSED** [review]

File: ddtrace/appsec/_asm_request_context.py [hunk 2]
```diff
@@ -128,7 +152,7 @@ def call_waf_callback(custom_data=None):
 
 def set_ip(ip):  # type: (Optional[str]) -> None
     if ip is not None:
-        set_value("waf_addresses", SPAN_DATA_NAMES.REQUEST_HTTP_IP, ip)
+        set_waf_address(SPAN_DATA_NAMES.REQUEST_HTTP_IP, ip, _ASM.get().span)
```
Comment (majorgreys) [inline ddtrace/appsec/_asm_request_context.py:155]: Can you explain why the change requires passing the asm request context's span? Why can't this be accomplished when the `_ASM.get()` is called to get the asm context in `set_value()`?

File: ddtrace/appsec/_asm_request_context.py [hunk 3]
```diff
@@ -94,6 +108,12 @@ def set_value(category, address, value):  # type: (str, str, Any) -> None
         asm_context_attr[address] = value
 
 
+def set_waf_address(address, value, span):  # type: (str, Any, Any) -> None
+    set_value("waf_addresses", address, value)
```
Comment (majorgreys) [inline ddtrace/appsec/_asm_request_context.py:112]: minor: use a private constant for the object attribute (category)

File: tests/tracer/test_trace_utils.py [hunk 4]
```diff
@@ -414,7 +418,7 @@ def test_set_http_meta(
             tag = "http.request.headers." + header
             assert span.get_tag(tag) == value
 
-    if appsec_enabled:
+    if appsec_enabled and span.span_type == SpanTypes.WEB:
```
Comment (majorgreys) [inline tests/tracer/test_trace_utils.py:421]: Should this bug fix be pulled out separately and backported?

Comment (christophe-papazian) [reply tests/tracer/test_trace_utils.py:421]: it did not have any impact previously as we were rewriting context at the end of the span. That's why it was not detected.

File: ddtrace/appsec/_asm_request_context.py (hunk 2 above)
Comment (christophe-papazian) [reply ddtrace/appsec/_asm_request_context.py:155]: Because now, we set the span context at the same time for metrics and tags purpose.

File: ddtrace/appsec/_asm_request_context.py (hunk 3 above)
Comment (christophe-papazian) [reply ddtrace/appsec/_asm_request_context.py]: Yes, this is done for all string literals now.

File: tests/tracer/test_trace_utils.py (hunk 4 above)
Comment (christophe-papazian) [reply tests/tracer/test_trace_utils.py:421]: But I tried to optimize things better so we are not rewriting data at the end of the span if it was already done by a previous call to the WAF. That's why it was detected only now.
